insights/v1alpha1/types_insights.go

```diff
@@ -0,0 +1,247 @@
+	// Data gathering is completed
+	Completed DataGatherState = "Completed"
+	// Data gathering failed
+	Failed DataGatherState = "Failed"
+	// Data gathering is pending
+	Pending DataGatherState = "Pending"
+	// Gatherer state marked as disabled, which means that the gatherer will not run.
+	Disabled GathererState = "Disabled"
+	// Gatherer state marked as enabled, which means that the gatherer will run.
+	Enabled GathererState = "Enabled"
+)
+
+// dataPolicy declares valid data policy types
+// +kubebuilder:validation:Enum="";ClearText;ObfuscateNetworking
+type DataPolicy string
+
+// state declares valid gatherer state types.
+// +kubebuilder:validation:Enum="";Enabled;Disabled
+type GathererState string
+
+// gathererConfig allows to configure specific gatherers
+type GathererConfig struct {
+	// name is the name of specific gatherer
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// state allows you to configure specific gatherer. Valid values are "Enabled", "Disabled" and omitted.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default.
+	// The current default is Enabled.
+	// +optional
+	State GathererState `json:"state"`
+}
+
+// dataGatherState declares valid gathering state types
+// +kubebuilder:validation:Optional
+// +kubebuilder:validation:Enum=Running;Completed;Failed;Pending
+// +kubebuilder:validation:XValidation:rule="!(oldSelf == 'Running' && self == 'Pending') && !(oldSelf == 'Completed' && self == 'Pending') && !(oldSelf == 'Failed' && self == 'Pending') && !(oldSelf == 'Completed' && self == 'Running') && !(oldSelf == 'Failed' && self == 'Running')", message="state cannot be changed backwards"
+type DataGatherState string
+
+// +kubebuilder:validation:XValidation:rule="(!has(oldSelf.insightsRequestID) || has(self.insightsRequestID)) && (!has(oldSelf.startTime) || has(self.startTime)) && (!has(oldSelf.finishTime) || has(self.finishTime)) && (!has(oldSelf.dataGatherState) || has(self.dataGatherState))",message="cannot remove attributes from status"
+// +kubebuilder:validation:Optional
+type DataGatherStatus struct {
+	// conditions provide details on the status of the gatherer job.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions" patchStrategy:"merge" patchMergeKey:"type"`
+	// dataGatherState reflects the current state of the data gathering process.
+	// +optional
+	State DataGatherState `json:"dataGatherState,omitempty"`
+	// gatherers is a list of active gatherers (and their statuses) in the last gathering.
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	Gatherers []GathererStatus `json:"gatherers,omitempty"`
+	// startTime is the time when Insights data gathering started.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="startTime is immutable once set"
```

Comment:
What's the zero value here, do you want to be able to change this form the zero value? Typically with a string we would include `oldSelf == '' ||` at the beginning of this

Reply:
Yes I think we want to change this from the zero value. Adding `oldSelf == '' || ` doesn't work though.  I am getting:
```
The CustomResourceDefinition "datagathers.insights.openshift.io" is invalid: spec.validation.openAPIV3Schema.properties[status].properties[finishTime].x-kubernetes-validations[0].rule: Invalid value: apiextensions.ValidationRule{Rule:"oldSelf == '' || self == oldSelf", Message:"finishTime is immutable once set"}: compilation failed: ERROR: <input>:1:9: found no matching overload for '_==_' applied to '(timestamp, string)'
 | oldSelf == '' || self == oldSelf
 | ........^

```

Reply:
If this doesn't work for you, I'm willing to adjust afterwards.  I suspect you need a cast of some kind.